README


                                CLORB

                a Common Lisp implementation of CORBA

                    Version 0.6 (an ALPHA release)

                          by Lennart Staflin

                              2005-05-05



CLORB is an Object Request Broker implementing CORBA 2. It currently
supports IIOP 1.0, IDL, DII, DSI, POA and value types. The goal is to
make the mapping follow the proposed mapping for LISP [1].

New in 0.6:

   * IIOP 1.1
   * ASDF support
   * IDL compiler improvments (fixed literals and constants, escaped chars
     in string literals) 

It lacks:

   * wide strings, chars
   * Request context
   * DynAny
   * CORBA Messaging
   * Multi-threading
   * IIOP 1.2

CLORB is released under GNU Library General Public License see the
file COPYING.


Ports:
   MCL 5        only tested with Max OS X native
   OpenMCL      working (Mac OS X)
   SBCL         working on Mac OS X and Debian GNU/Linux (i386)
   CLISP        works again (2.32)
   CMUCL        working with 19a on Mac OS X and Debian GNU/Linux (i386)

Non-working ports:
   ACL 6.2      



[1] http://www.omg.org/technology/documents/formal/lisp_language_mapping.htm



                     Compiling and Loading CLORB

The file clorb-files.lisp contains code to compile and load CLORB.
There is also an ASDF system definition file clorb.asd. The ASDF
definition is good enough for loading CLORB, but inconvenient for
hacking.

clorb-files.lisp defines a package NET.CDDR.CLORB.SYSTEM and a
function RELOAD (in that package) that will compile and load CLORB.
It will load files relative to the directory from where clorb-files
was loaded. 

The files:

- devel.lisp - Loads CLORB for development purpose, it also initiates
  the ORB and the object adaptor. It runs all the test cases and
  prepares the "hello world" example (use (hh) to run it or (hhn) to
  use the Name Service).

- loadup.lisp - Loads CLORB for normal use. The code will be loaded
  but not initialized.

It is not necessary to setup any logical pathname hosts to load
CLORB. But some extra features like test cases and examples might
depend on the "CLORB" host. It should then be set up to map
CLORB:SRC; to the main CLORB sources and allow relative directory
references.

When compiling and loading clorb some clorb specific features control
the process. 

    
Features to control what socket/tcp code gets used:

    :use-acl-socket     use the ACL-SOCKET package in preference to
                        native socket support (requires the acl
                        compatibility package)

    :dummy-tcp          don't include any socket code (can be used to
                        test the rest of the code)


Features that are automatically determined (in clorb-sysdep):

    clorb::cmucl-sockets
    clorb::db-sockets
    clorb::sb-bsd-sockets
    clorb::mcl-bsd


                         Initializing the ORB


After CLORB has been loaded and before the ORB can be used it has to
be initialized. This is done with the CORBA:ORB_init function.

(CORBA:ORB_init arguments orb-id) => orb-object, arguments'

Where arguments is a list of strings (as would usually be passed to
a command line program in Unix) and orb-id is a string identifying
the particular orb (only one is supported pass "" for this). The
function can be called more than once, and will return the same
object. The argument list will be processed by every call.

In the arguments list, items starting with "-ORB" will be recognized
and removed (the item after will also be removed if it provides
argument to the -ORB argument). The remaining arguments will be
returned as a second return value. The recognized items are:

    -ORBInitRef Name=IOR

      This provides value for the initial references. A call to
      op:resolve_initial_references with Name will return the object
      references by the IOR.

    -ORBDefaultInitRef IOR

    -ORBPort port#

      Specify the port that the Object adaptor will listen on.

    -ORBHostname name

      Specify the hostname that will be included in object references,
      could also be an IP-number.


The clorb:*log-level* controls the amount of logging done by CLORB.
It should be an integer, lower value gives more logging. Logging is
written to the stream *log-output*.



                     Loading/Compiling IDL files

(CORBA:IDL pathname &key eval print output skeleton) => repository

Load and parse the IDL file referenced by PATHNAME. The interface
repository containing the parsed IDL file is returned. The necessary
code to use this interface is generated and loaded (unless the eval
key is nil).

Producing a Lisp file with all code for an IDL file:

  (CORBA:IDL "foo.idl" :output "foo.lisp")

The abow will also load the code. To only produce the file specify
:eval nil. (Some packages might still be created as a side effect.)



                 Obtaining initial object references


(op:list_initial_services orb) => list-of-names

(op:resolve_initial_references orb "NameService") => object-reference

(op:string_to_object orb "IOR") => object-reference

The IOR could be on the IOR:xxx form or it could be a file or http
URL, in that case the references resource is should contain an IOR.
The IOR could also be a corbaloc or corbaname IOR.




                           Client Operation

Initialize and get a reference to the ORB object.

    (defvar *the-orb* (CORBA:ORB_init))

Load the IDL for the interface.

    (corba:idl "clorb:idl;Random.idl")

Obtain an object reference and call methods on the reference.

    (let ((obj
           (op:string_to_object *the-orb* "http://www.random.org/Random.ior")))
      (op:lrand48 obj))


[Unfortunately the CORBA server for random number where not answering
the last time I tried it.]



                           Server Operation

Initialize the ORB and load the IDL for the interface, as for Client
Operation.

Get a reference to the Root POA (Object Adapter).

    (defvar *poa* (op:resolve_initial_references *orb* "RootPOA"))

Possibly create another POA with policies tailored for the server
needs.

Activate the POA:

    (op:activate (op:the_poamanager *poa*))

Create and register a servant:

    (defvar *servant* (make-instance 'my-class))
    (defvar *object* (op:activate_object *poa* *servant*))

Get stringified IOR for object

    (op:object_to_string *orb* *object*)

The string can be transfered to the client in whatever way.
Or register with name service

   (clorb:rebind *object* "my-name")


                                MCL 5

MCL 5.0 is the current main development environment. You also need the
BSD package <URL:http://babs.cs.umass.edu/~bburns/software/bsd.lisp>.
The networking code can either use OpenTransport directly or use the
ACL-COMPAT code from portable allegro server [fix reference].

To load CLORB you need to do something like this:

    (require 'bsd)
    (load "CLORB:SRC;CLORB-FILES")
    (net.cddr.clorb.system:reload)



                               OpenMCL

(load "clorb-files")
(net.cddr.clorb.system:reload)

;; Initializing the ORB
(defvar *orb* 
  (CORBA:ORB_init
   (list ;;"-ORBPort" "4712"
         ;;"-ORBInitRef" "NameService=corbaloc::127.0.0.1:4711/NameService"
         )))


                                 SBCL

Tried with version 0.9.0 on MacOS X, 0.8.16 on Debian GNU/Linux i386.

The system dependent part will (require :sb-bsd-sockets).

(load "clorb-files")
(net.cddr.clorb.system:reload)

;; Initializing the ORB
(defvar *orb* 
  (CORBA:ORB_init
   (list ;;"-ORBPort" "4712"
         ;;"-ORBInitRef" "NameService=corbaloc::127.0.0.1:4711/NameService"
         )))


                                CLISP

You might need a recent version of CLISP for CLORB to work. I have
tested with 2.33 on both Debian GNU/Linux (i386) and on Mac OS X 10.2.8.

There is a pretty printer bug in CLISP 2.32 that cause problems for
the CORBA:IDL function with either :print or :output options. You can
get around this by also specifying :pprint-dispatch nil. (This seems
to be needed for 2.33 also.)


                         About the CMUCL port

CLORB 0.6 seems to work with CMUCL 19a. Except for a problem loading
the fasl file for cosnaming-stub. A work-around is to

    (setf c::top-level-lambda-max 0)

before compiling.

Tested on Debian GNU/Linux (i386)
and Mac OS 10.3.9.



From: Daniel Barlow <dan@telent.net>
Date: 11 May 2000 00:39:38 +0100

[...]

2) The user has a choice of socket libraries: the standard CMUCL
socket code, or my replacement sockets library.  The standard code
doesn't support the socket option SO_REUSEADDR (without writing ffi
glue) so you need my sockets if you want to do server applications.
You can get it at http://ww.telent.net/lisp/sockets.html

3) The system is built using mk-defsystem, which comes with CMUCL
(if you use the Debian packages, at any rate).

4) Only tested on ix86 GNU/Linux so far.

5) It works, kind of.  It's not been extensively tested.  I can make
it talk as a client to the GNOME Help viewer (see
http://ww.telent.net/corba/gnome-lisp.html for a worked example), and
I can make the ORBit 'name-client' test program talk to your CosNaming
service implementation.  Serving is a bit CPU-intensive until I get it
actually blocking instead of busy-waiting when waiting for a connection.

6) It's not been tested recently in SBCL (Steel Bank Common Lisp) but
there's no particularly good reason for it not to work there too.
You'll have to use my socket library in that one, because it doesn't
have any socket code of its own at all.

[...]

                               Examples

The subdirectory examples contain some examples.

1. Hello World

In examples/hello. This is a simple client/server example. The file
hello.idl describes the interface:

module CLORB_EX {
  interface HelloWorld {
    string greet ();
  };
};

The server implements a servant with a method "greet" that returns a
string. The client obtains an object reference and calls the greet
method.

Load examples/hello/loadup (loaded by default by devel.lisp), and the
in the server run

(run-hello :file "/tmp/hello.ior")


In the client run

(hello-client :file "/tmp/hello.ior")


The :file argument specifies how the object reference is exchanged.
If you have a name server configured you can use then :name argument
instead.




                              Test Suite

CLORB comes with a test suite. The test coverage is currently spotty.

The support system for the test cases has now been put in its own
package. I have also used it for some other projects. The code is
however included in the luna subdirectory. Some CLORB specific
support is in the support-test.lisp. 

The files test-*.lisp contains test suits for parts of CLORB and the
file all-test.lisp runs all the test suites and aggreates the result.

                       About the included files

clorb-*.lisp

This is main part of the ORB.

clorb-options.lisp

Some parameters that might need to be customized. Especially
host-name and how to find the Naming Service and Interface
Repository.


pns-server.lisp

A new implementation of CosNaming, where the state is stored
in files (thus persistent).


idef-read.lisp
idef-write.lisp

Read and write IDEF, an sexp version of IDL. These reads and
writes from the local Interface Repository implemention.

idef-macros.lisp
Macro wrapper around idef-read to allow IDEF to be loaded from
file. Also macro for defining servant skeleton class for use
with the CLORB auto-servant class.


cosevent-idl.lisp
cosnaming-idl.lisp

IDEF version of standard IDL for CosEventChannel and
CosNaming. (generated by idef-write)


cosnaming-stub.lisp

The types, classes for CosNaming according to the OMG Lisp Mapping.
Proxy classes and operation methods.
(This partly hand created static stub.)

cosnaming-skel.lisp

Servant classes for CosNaming.


orbit.lisp

Set *principal* for communication with ORBit (GNOME) servers.


EXAMPLE CODE (in examples)

examples/hello/

Hello World example, autoloaded by devel.lisp (via examples/hello/auto.lisp)
Try: (hh) or (hhn)
Or, in server: (run-hello :name "hello")
in client:     (hello-client :name "hello")
assuming you have a running name server.


examples/dii-example.lisp
examples/dsi-example.lisp 
examples/dsi/clive.lisp

Example of using DSI (dynamic servant implementation) and DII
(dynamic invokation interface) using standard interfaces.
From a discussion on comp.lang.corba





ChangeLog
2009-11-15  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-util.lisp (print-escape-name): new fun
	(to-string): new fun
	(rebind): added simple restart to retry binding

2008-12-26  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-value.lisp (unmarshal-value-tag): renamed from
	unmarshal-chunk-tag, and :chunk-tag renamed to :value-tag
	(marshal-record): avoid padding current chunk when starting new value

2008-12-25  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-ifr.lisp (idltype-tc): union-def labels any-value, not anys

	* clorb-buffer.lisp (align-pos): new func
	(begin-in-buffer): new func
	(with-in-buffer): use begin-in-buffer for align - handle padded chunks

2008-10-04  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-poa.lisp (poa-dispatch): new minor code to CORBA:OBJECT_NOT_EXIST
	(poa-locate): return nil if poa not found

2008-09-15  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-conn.lisp (print-object): fix bug in printing connections.

2008-08-07  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-supp.lisp (*log-writer*): new var
	(clorb-simple-log-writer): new fun
	(mess): use *log-writer*

2008-08-03  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-srv.lisp (connection-receive-request): remove principal
	from logging

2008-08-02  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-util.lisp (orb-resolve): accept namestr=nil, to return
	namecontext instead of resolving name.

	* clorb-orb.lisp (corbaname-to-object): call orb-resolve with nil
	name-part if name-part is missing.

2008-07-25  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-request.lisp (request-send): also send locate-request if
	op is locate
	(request-poll): handle object_forward for locate-reply
	(request-get-response): handle respons from locate-reply

	* clorb-iiop.lisp (connection-reply): use object_forward not
	location_forward for locatereply
	(locate): shouldn't be used

	* clorb-any.lisp (create-any): new func
	(corba:any): use new func
	(corba:any): compiler-macro

	* clorb-util.lisp (send): new func
	(results): new func
	(poll): new func
	(op::send_deferred, op::send_oneway): methods for string op as
	first arg
	(op::send_deferred, op::send_oneway): methods typecode as first arg

2008-07-13  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-orb.lisp (pull): renamed from poll

2008-07-12  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-util.lisp (create-request): renamed from object-create-request
	(create-request): change arg order, support new request-arglist
	(create-request-noir): new func
	(create-request-ir): new func

2008-05-12  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-io.lisp (io-descriptor): add peer slot

2008-05-09  Lennart Staflin  <lenst@lysator.liu.se>

	Patch from Ilya Zvyagin:

	* clorb-sysdep.lisp (shell-to-stream): support for win32 SBCL
	(cpp-command-string): use Microsoft Visual C Compiler preprocessor
	on win32.

	* clorb-idlcpp.lisp (parse-cpp-line): Microsoft VC C/C++ compiler support

2007-02-27  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-macros.lisp (opname-helper): ORG/FEATURE typo fix.

	* clorb-pkgdcl.lisp ("OMG.ORG/FEATURE"): renames from ORG/FEATURES
	(typo).

2006-07-11  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-util.lisp (op:bind): new
	(op:rebind): new

2006-07-07  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-servant.lisp (request-respond): fix exception handling

2006-07-02  Lennart Staflin  <lenst@lysator.liu.se>

	* test-poa.lisp (mock-dsi-servant): new plug-invoke slot
	(mock-dsi-default-invoke): new
	(invoke): use new plug-invoke slot
	("POA Test"): new "Async Servant" test

	* clorb-servant.lisp (request-respond): changed name from
	server-request-respond
	(request-respond): exception handling fixed

	* clorb-poa.lisp (make-poa-current): capture request and servant instead
	(poa-current-request): new
	(current-request): new, for easy access
	(poa-invoke): set the poa slot
	(poa-invoke): new make-poa-current interface

	* clorb-io.lisp (io-system-add-descriptor, io-system-delete-descriptor):
	added ignore desc


2006-06-09  Lennart Staflin  <lenst@lysator.liu.se>

	* Version 0.7

2006-06-08  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-util.lisp (object-create-request): use static-error-handler

	* clorb-request.lisp (client-request): use lazy synchronization
	(client-request): added slot status-notify-callback
	(status-notify): new generic function, and method for client-request
	(*send-hook*): new hook
	(request-funcall): corba:funcall will now use input-func that
	returns result
	(dii-input-func): return values

	* clorb-orb.lisp (bg-run): handle case of missing adapter

	* clorb-mt.lisp (synchronized-lazy): new class

2006-05-19  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-io.lisp (io-descriptor-connected-p): spelling

	* clorb-srv.lisp (poa-connection-handler): use new :server-p arg

	* clorb-conn.lisp (make-associated-connection): added &key server-p

	* clorb-io.lisp: removed *io-descriptions*, io-fix-broken,
	class io-system-multiprocess and related methods
	(io-system): added slot io-descriptions
	(io-system-reset io-system): does the kill processes and close
	streams stuff.
	(io-system-add-descriptor): new
	(io-system-delete-descriptor): new
	(io-descriptor-destroy): added &optional soft - to check that no
	writer is using the descriptor
	(io-desciptor-connected-p): new
	(io-broken-descriptor): added &optional activity to close the
	socket if possible

2006-05-18  Lennart Staflin  <lenst@lysator.liu.se>

	* clorb-sysdep.lisp (socket-close): new

	* clorb-srv.lisp (connection-receive-request): check
	shutdown-status, and ingnore request if connection is shuting down

	* clorb-request.lisp (request-no-write): new

	* clorb-io.lisp (io-descriptor-destroy): use socket-close
	(%io-descriptor-close): new
	(io-descriptor-close): new
	(io-descriptor-set-write): if descriptor broken generate a :no-write event
	(io-broken-descriptor): send no event
	(io-write-broken-descriptor): new

devel.lisp
(in-package :cl-user)

(format t ";;; Loading CLORB~%")


;;; MCL 5.0 - needs the BSD package

#+digitool
(unless (find-package "BSD")
  (cond ((find-package "NET.CDDR.PACKER")
         (funcall (intern "REQUIRE-PACKAGE" "NET.CDDR.PACKER") "BSD"))
        (t
         (require "BSD"))))


;;; Load the CLORB system

(load (merge-pathnames "clorb-files.lisp" *load-pathname*))
(net.cddr.clorb.system:set-load-opts
 :server t  :idlcomp nil  :my-idlparser t
 :portable-interceptor t  :support-test t
 :hello-world t)
(net.cddr.clorb.system:reload)


#+digitool
(import '(@ @@ @@@) "CLORB")

;; To easy switch between CLORB and CL-USER from the listener
(define-symbol-macro inpc (in-package :clorb))
(define-symbol-macro clorb::inpu (in-package :cl-user))


;;; SBCL repl support, runs the server parts using add-fd-handler
#+sbcl
(clorb::make-repl-happy)


;;; IO Policy

(setq clorb:*io-system-default-class*
      (or #+(or mcl openmcl) 'clorb:io-system-mt-blocking-write
          'clorb:io-system-select-blocking-write))

;;; Initiating the ORB
;; provide initial references to services on the system

(defvar *the-orb*
  (CORBA:ORB_init
   (list #+digitool "-ORBPort 4711"
         ;;#+cmu      "-ORBPort 4713"
         ;;"-ORBInitRef NameService=corbaloc::/NameService" 
         "-ORBInitRef NameService=corbaloc:rir:/CLORB-PNS" )))


;;; Initiating the object adapter (the server part of the ORB)

(format t "~&;;; Activating the POA~%")
(op:activate (op:the_poamanager (clorb::root-poa)))


;;; Examples
;;
;;(load (merge-pathnames
;;       (make-pathname :name "auto" :type "lisp"
;;                      :directory '(:relative "examples" "hello"))
;;       clorb::*clorb-pathname-defaults*))
;;
;; run hello world with (hh)
;; or (hhn) using the name service


;;; Run test cases
#-clorb-no-test
(load (merge-pathnames "all-test.lisp" clorb::*clorb-pathname-defaults*))
